File: 201-300/q206_reverse_linked_list.java
// =========== 2 solutions =============

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// iterative solution
class Solution {
    public ListNode reverseList(ListNode head) {
        
        ListNode curr = head;
        ListNode prev = null;
        ListNode next = null;
        
        while(null != curr)
        {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        
        return prev;
    }
}

// recursive solution
class Solution {
    public ListNode reverseList(ListNode head) {
        
        if(null == head || null == head.next)
        {
            return head;
        }
        
        ListNode temp =  reverseList(head.next);
        head.next.next = head;
        head.next = null;
        
        return temp;
    }
}

File: 201-300/q206_reverse_linked_list.c
// =========== 2 solutions =============

//excercise struct
struct ListNode {
    int val;
    struct ListNode *next;
};
 

/*iterative*/
struct ListNode* reverseListIter(struct ListNode* head){

    struct ListNode *curr = head;
    struct ListNode *next = NULL;
    struct ListNode *prev = NULL;
    
    while(NULL != curr)
    {
        next = curr->next;
        curr->next = prev;
        prev = curr;
        curr = next;
    }
    
    return prev;
}

/*recursive*/
struct ListNode* reverseListRecursive(struct ListNode* head){

    if(NULL == head || NULL == head->next)
    {
        return head;
    }
    
    struct ListNode *temp = reverseListRecursive(head->next);
    head->next->next = head;
    head->next = NULL;
    
    return temp;
}